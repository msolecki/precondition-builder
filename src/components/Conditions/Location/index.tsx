import React, {ChangeEvent} from 'react'
import {makeStyles} from '@material-ui/core/styles'
import InputLabel from '@material-ui/core/InputLabel'
import FormControl from '@material-ui/core/FormControl'
import InputAdornment from '@material-ui/core/InputAdornment'
import {Input} from '@material-ui/core'
import PropTypes from 'prop-types'
import Grid from '@material-ui/core/Grid'
import Title from '../../Title'

const useStyles = makeStyles(({spacing}) => ({
    root: {
        display: 'flex',
        flexWrap: 'wrap',
    },
    formControl: {
        marginTop: spacing(1),
        marginBottom: spacing(3),
        minWidth: 120,
    },
    formControlWithoutMargin: {
        marginTop: spacing(1),
        minWidth: 120,
    },
}))

export interface LocationProps {
    setRadius(value: number): void;

    setLatLng(value: string): void;

    radius: number | null;
    latLng: string | null;
}

const Location: React.FC<LocationProps> = (props: LocationProps): React.ReactElement => {
    const classes = useStyles()
    const [latLngError, setLatLngError] = React.useState<boolean>(false)

    function handleRadiusChange(event: ChangeEvent<{}>): void {
        const {value} = event.target as HTMLInputElement

        props.setRadius(Number(value))
    }

    function isValidatedLatLangFormat(value: string): boolean {
        return /^@([-+]?\d{1,2}[.]\d+),([-+]?\d{1,2}[.]\d+)$/.test(value)
    }

    function handleLatLangChange(event: ChangeEvent<{}>): void {
        const {value} = event.target as HTMLInputElement
        const validated = isValidatedLatLangFormat(value)

        setLatLngError(!validated)

        props.setLatLng(value)
    }

    return (
        <>
            <Title text="Location"/>
            <Grid container spacing={3}>
                <Grid item xs={6}>
                    <FormControl fullWidth required className={classes.formControlWithoutMargin} component='div'>
                        <InputLabel htmlFor="location-radius">Radius</InputLabel>
                        <Input
                            id="location-radius"
                            placeholder={'0'}
                            type='number'
                            value={Number(props.radius) || ''}
                            onChange={handleRadiusChange}
                            endAdornment={<InputAdornment position="end">m</InputAdornment>}
                            inputProps={{min: 5, max: 10000, step: 1}}
                        />
                    </FormControl>
                </Grid>
                <Grid item xs={6}>
                    <FormControl fullWidth required className={classes.formControl} component='div'>
                        <InputLabel htmlFor="location-latlng" error={latLngError}>GPS</InputLabel>
                        <Input
                            id="location-gps"
                            value={props.latLng || ''}
                            placeholder={'@52.13,13.45'}
                            onChange={handleLatLangChange}
                            error={latLngError}
                        />
                    </FormControl>
                </Grid>
            </Grid>
        </>
    )
}

Location.propTypes = {
    setRadius: PropTypes.func.isRequired,
    setLatLng: PropTypes.func.isRequired,

    radius: PropTypes.number,
    latLng: PropTypes.string,
}

Location.defaultProps = {
    radius: null,
    latLng: null,
}

export default Location
