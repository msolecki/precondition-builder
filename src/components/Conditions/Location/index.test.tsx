import * as React from 'react'
import {ReactTestRenderer} from 'react-test-renderer'
import {createShallow} from '@material-ui/core/test-utils'

import Location, {LocationProps} from './index'

const component = (props: LocationProps): Location => <Location {...props}/>
const setRadiusFunc = jest.fn()
const setLatLngFunc = jest.fn()

describe('<Location />', () => {
    let shallow

    beforeAll(() => {
        shallow = createShallow()
    })

    test('empty snapshot', () => {
        const wrapper: ReactTestRenderer = shallow(
            component({
                setRadius: () => {
                },
                setLatLng: () => {
                },
                radius: null,
                latLng: null,
            }))

        expect(wrapper).toMatchSnapshot()
    })

    test('default snapshot', () => {
        const wrapper = shallow(
            component({
                setRadius: setRadiusFunc,
                setLatLng: setLatLngFunc,
                radius: 123,
                latLng: 'test',
            }))
        const eventRadius = {target: {name: 'name', value: 123}}
        const eventLatLng = {target: {name: 'name', value: 'test'}}

        wrapper.find('#location-radius').simulate('change', eventRadius)
        wrapper.find('#location-gps').simulate('change', eventLatLng)

        expect(setRadiusFunc).toHaveBeenCalledWith(123)
        expect(setLatLngFunc).toHaveBeenCalledWith('test')
    })
})
