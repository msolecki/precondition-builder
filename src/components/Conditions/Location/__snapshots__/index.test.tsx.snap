// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Location /> empty snapshot 1`] = `
<Fragment>
  <Title
    text="Location"
  />
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={6}
    >
      <WithStyles(ForwardRef(FormControl))
        className="makeStyles-formControlWithoutMargin-3"
        component="div"
        fullWidth={true}
        required={true}
      >
        <WithStyles(WithFormControlContext(ForwardRef(InputLabel)))
          htmlFor="location-radius"
        >
          Radius
        </WithStyles(WithFormControlContext(ForwardRef(InputLabel)))>
        <WithStyles(ForwardRef(Input))
          endAdornment={
            <ForwardRef(WithStyles)
              position="end"
            >
              m
            </ForwardRef(WithStyles)>
          }
          id="location-radius"
          inputProps={
            Object {
              "max": 10000,
              "min": 5,
              "step": 1,
            }
          }
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </WithStyles(ForwardRef(FormControl))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={6}
    >
      <WithStyles(ForwardRef(FormControl))
        className="makeStyles-formControl-2"
        component="div"
        fullWidth={true}
        required={true}
      >
        <WithStyles(WithFormControlContext(ForwardRef(InputLabel)))
          error={false}
          htmlFor="location-latlng"
        >
          GPS
        </WithStyles(WithFormControlContext(ForwardRef(InputLabel)))>
        <WithStyles(ForwardRef(Input))
          error={false}
          id="location-gps"
          onChange={[Function]}
          placeholder="@52.13,13.45"
          value=""
        />
      </WithStyles(ForwardRef(FormControl))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Fragment>
`;
